File: lab04/analyzer.py
#!/usr/bin/env python3
########################################
# File: analyzer.py
#
#   ------------------------------------
#   Lab 04: Time Complexity & Profiling
#   ------------------------------------
#
#   Author: Daniel Li
#   Course: COP3504C
#   Section: 25452 (Tavassoli)
#   Date: Sep 12 2024
#
########################################

from time import time

# Students will implement the analyzer module with a main entry point
# and two search methods as follows.

#---------------------------------------
# Search Methods
#---------------------------------------

# linear_search(container: tuple(str), element: str) -> int
#
#   Uses linear search to find specified element in container.
#   Returns index of element, or -1 if not found.
#
def linear_search(container, element):
    for index in range(len(container)):
        if container[index] == element:
            return index
    return -1

# binary_search(container: tuple(str), element: str) -> int
#
#   Uses binary search to find specified element in container.
#   Returns index of element, or -1 if not found.
#
def binary_search(container, element):
    lower = 0
    upper = len(container)
    while lower < upper:
        mid = (lower + upper) // 2
        if element < container[mid]:
            upper = mid
        elif element > container[mid]:
            lower = mid + 1
        else:
            return mid
    return -1

#---------------------------------------
# Other Methods
#---------------------------------------

# This function tests both search methods by searching for
# the specified element in a list.
def profile_search(container, element):
    print(f'Searching for \'{element}\':')

    start_time = time()
    index = linear_search(container, element)
    end_time = time()
    delta_time = end_time - start_time
    print(f'  linear search: found {index} in {delta_time}s')

    start_time = time()
    index = binary_search(container, element)
    end_time = time()
    delta_time = end_time - start_time
    print(f'  binary search: found {index} in {delta_time}s')

#---------------------------------------
# Entry Point (main)
#---------------------------------------

# This module will have a main() function which will execute if the module is
# directly invoked. The main function should complete the following steps,
# documenting results and answer questions in the write-up:
#
#     1) Access the data set from the stringdata module (via the dataset value).
#     2) Search for "not_here" using both algorithms. Capture the time each method requires to execute.
#     3) Search for "mzzzz" using both algorithms. Capture the time each method requires to execute.
#     4) Search for "aaaaa" using both algorithms. Capture the time each method requires to execute.
#
def main():
    import stringdata
    print('Generating dataset...')
    start_time = time()
    dataset = stringdata.get_data()
    end_time = time()
    delta_time = end_time - start_time
    print(f'Generated {len(dataset)} values in {delta_time}s')

    profile_search(dataset, 'not_here')
    profile_search(dataset, 'mzzzz')
    profile_search(dataset, 'aaaaa')

if __name__ == '__main__':
    main()

File: lab05/analyzer.cpp
/***************************************
 * File: analyzer.py
 *
 *  ------------------------------------
 *  Lab 05: Software Engineering
 *  ------------------------------------
 *
 *  Author: Daniel Li
 *  Course: COP3504C
 *  Section: 25452 (Tavassoli)
 *  Date: Oct 01 2024
 *
 **************************************/

#include <chrono>
#include <iostream>
#include <vector>

//--------------------------------------
// Search Methods
//--------------------------------------

// Uses linear search to find specified element in container.
// Returns index of element, or -1 if not found.
int linear_search(const std::vector<std::string>& container, const std::string& element)
{
  for (unsigned int i = 0; i < container.size(); i++)
    if (container[i] == element)
      return i;
  return -1;
}

// Uses binary search to find specified element in container.
// Returns index of element, or -1 if not found.
int binary_search(const std::vector<std::string>& container, const std::string& element)
{
  int lower = 0;
  int upper = container.size();
  while (lower < upper) {
    int mid = (lower + upper) / 2;
    int cmp = element.compare(container[mid]);
    if (cmp < 0)
      upper = mid;
    else if (cmp > 0)
      lower = mid + 1;
    else
      return mid;
  }
  return -1;
}

//--------------------------------------
// Other Methods
//--------------------------------------

//
// A simplified data generator that I made.
//
std::vector<std::string> generate_data()
{
  char str[6];
  int i = 0;
  str[5] = '\0';

  std::vector<std::string> dataset(26*26*26*26*26);
  for (str[0] = 'a'; str[0] <= 'z'; str[0]++)
    for (str[1] = 'a'; str[1] <= 'z'; str[1]++)
      for (str[2] = 'a'; str[2] <= 'z'; str[2]++)
        for (str[3] = 'a'; str[3] <= 'z'; str[3]++)
          for (str[4] = 'a'; str[4] <= 'z'; str[4]++)
            dataset[i++] = str;
  return dataset;
}

//
// This function tests both search methods.
//
void profile_search(const std::vector<std::string>& container, const std::string& element)
{
  typename std::chrono::system_clock::time_point start_time, end_time;
  std::chrono::duration<double, std::milli> delta_time;
  int index;

  std::cout << "Searching for '" << element << "':" << std::endl;

  start_time = std::chrono::system_clock::now();
  index = linear_search(container, element);
  end_time = std::chrono::system_clock::now();
  delta_time = std::chrono::duration_cast<decltype(delta_time)>(end_time - start_time);
  std::cout << "  linear search: found " << index << " in " << delta_time.count() << "ms" << std::endl;

  start_time = std::chrono::system_clock::now();
  index = binary_search(container, element);
  end_time = std::chrono::system_clock::now();
  delta_time = std::chrono::duration_cast<decltype(delta_time)>(end_time - start_time);
  std::cout << "  binary search: found " << index << " in " << delta_time.count() << "ms" << std::endl;
}

//--------------------------------------
// Entry Point (main)
//--------------------------------------

int main()
{
  typename std::chrono::system_clock::time_point start_time, end_time;
  std::chrono::duration<double, std::milli> delta_time;

  std::cout << "Generating dataset..." << std::endl;
  start_time = std::chrono::system_clock::now();
  std::vector<std::string> dataset = generate_data();
  end_time = std::chrono::system_clock::now();
  delta_time = std::chrono::duration_cast<decltype(delta_time)>(end_time - start_time);
  std::cout << "Generated " << dataset.size() << " values in " << delta_time.count() << "ms" << std::endl;

  profile_search(dataset, "not_here");
  profile_search(dataset, "mzzzz");
  profile_search(dataset, "aaaaa");

  return 0;
}
